synapse/handlers/user_directory.py

```diff
@@ -277,44 +266,18 @@ def _handle_initial_room(self, room_id):
                 count += 1
 
                 user_set = (user_id, other_user_id)
```

Comment:
having said that, I'm not quite sure why we want entries for `(user_id, user_id)`. It feels like we have to go to special effort to add it, and then more special effort to remove it, and it would be simpler all round to avoid it in the first place?

(if we're going to avoid it, I'd advocate enforcing that with a database constraint if possible because you *know* it's going to leak in otherwise)

Reply:
okay, so I've removed `(user_id, user_id)` and fixed the sytests that rely on being able to search for yourself.

There is also the bug that by using the users_who_share_public_rooms table, a user who is in a public room with nobody else and no other rooms will not share a room with anybody and therefore will not show up in search. I think this is probably fine, though.